sdk/core/azure-core/inc/http/http.hpp

```diff
@@ -63,6 +66,191 @@ enum class HttpMethod
   PATCH,
 };
 
+class Request
+{
+
+private:
+  // query needs to be first or at least before url, since url might update it
+  std::map<std::string, std::string> _queryParameters;
+
+  HttpMethod _method;
+  std::string _url;
+  std::map<std::string, std::string> _headers;
+  std::map<std::string, std::string> _retryHeaders;
+  std::map<std::string, std::string> _retryQueryParameters;
+  // Request can contain no body, or either of next bodies (_bodyBuffer plus size or bodyStream)
+  BodyStream* _bodyStream;
+  BodyBuffer* _bodyBuffer;
+
+  // flag to know where to insert header
+  bool _retryModeEnabled;
+
+  // returns left map plus all items in right
+  // when duplicates, left items are preferred
+  static std::map<std::string, std::string> mergeMaps(
+      std::map<std::string, std::string> const& left,
+      std::map<std::string, std::string> const& right)
+  {
+    auto result = std::map<std::string, std::string>(left);
+    std::for_each(right.begin(), right.end(), [&](auto const& pair) { result.insert(pair); });
+    return result;
+  }
+
+  /**
+   * Create an stream from source string. Then use getline and separator to get as many tokens as
+   * possible and insert each one to result vector
+   */
+  static std::vector<std::string> split(std::string source, char separator)
+  {
+    auto result = std::vector<std::string>();
+
+    auto stringAsStream = std::stringstream(source);
+    auto token = std::string();
+
+    while (std::getline(stringAsStream, token, separator))
+    {
+      result.push_back(token);
+    }
+    return result;
+  }
+
+  /**
+   * Will check if there are any query parameter in url looking for symbol '?'
+   * If it is found, it will insert query parameters to _queryParameters internal field
+   * and remove it from url
+   */
+  std::string parseUrl(std::string url)
+  {
+    auto position = url.find('?');
```

Comment:
How about using iterators to avoid needing so many string copies? Something like this:

```
const auto firstPosition = std::find(url.begin(), url.end(), '?');
if (firstPosition == url.end()) { return url; }
auto position = firstPosition;
while (position != url.end()) {
    ++position; // skip over the ? or &
    const auto nextPosition = std::find(position, url.end(), '&');
    const auto equalChar = std::find(position, nextPosition, '=');
    auto valueStart = equalChar;
    if (valueStart != nextPosition) { ++valueStart; } // skip over the =, if any
    if (std::find(valueChar, nextPosition, '=') != nextPosition) { throw /* too many equals ? */; }
    _queryParameters.insert(std::string(position, equalChar), std::string(valueStart, nextPosition));
    position = nextPosition;
}

return std::string(url.begin(), firstPosition);
```

Reply:
beautiful.
Yeah, I am taking it (if you dont mind)
I will remove the validation for having many equal symbols in a parameter value (pretty sure we don't need that for now)